lib/vfs/src/lib.rs

```diff
@@ -0,0 +1,443 @@
+        conf: &OpenOptionsConfig,
+    ) -> Result<Box<dyn VirtualFile>, FsError>;
+}
+
+#[derive(Debug, Clone)]
+pub struct OpenOptionsConfig {
+    read: bool,
+    write: bool,
+    create_new: bool,
+    create: bool,
+    append: bool,
+    truncate: bool,
+}
+
+impl OpenOptionsConfig {
+    pub const fn read(&self) -> bool {
+        self.read
+    }
+
+    pub const fn write(&self) -> bool {
+        self.write
+    }
+
+    pub const fn create_new(&self) -> bool {
+        self.create_new
+    }
+
+    pub const fn create(&self) -> bool {
+        self.create
+    }
+
+    pub const fn append(&self) -> bool {
+        self.append
+    }
+
+    pub const fn truncate(&self) -> bool {
+        self.truncate
+    }
+}
+
+// TODO: manually implement debug
+
+pub struct OpenOptions {
+    opener: Box<dyn FileOpener>,
+    conf: OpenOptionsConfig,
+}
+
+impl OpenOptions {
+    pub fn new(opener: Box<dyn FileOpener>) -> Self {
+        Self {
+            opener,
+            conf: OpenOptionsConfig {
+                read: false,
+                write: false,
+                create_new: false,
+                create: false,
+                append: false,
+                truncate: false,
+            },
+        }
+    }
+
+    pub fn read(&mut self, read: bool) -> &mut Self {
+        self.conf.read = read;
+        self
+    }
+
+    pub fn write(&mut self, write: bool) -> &mut Self {
+        self.conf.write = write;
+        self
+    }
+
+    pub fn append(&mut self, append: bool) -> &mut Self {
+        self.conf.append = append;
+        self
+    }
+
+    pub fn truncate(&mut self, truncate: bool) -> &mut Self {
+        self.conf.truncate = truncate;
+        self
+    }
+
+    pub fn create(&mut self, create: bool) -> &mut Self {
+        self.conf.create = create;
+        self
+    }
+
+    pub fn create_new(&mut self, create_new: bool) -> &mut Self {
+        self.conf.create_new = create_new;
+        self
+    }
+
+    pub fn open<P: AsRef<Path>>(&mut self, path: P) -> Result<Box<dyn VirtualFile>, FsError> {
+        self.opener.open(path.as_ref(), &self.conf)
+    }
+}
+
+/// This trait relies on your file closing when it goes out of scope via `Drop`
+#[cfg_attr(feature = "enable-serde", typetag::serde)]
+pub trait VirtualFile: fmt::Debug + Send + Write + Read + Seek + 'static + Upcastable {
+    /// the last time the file was accessed in nanoseconds as a UNIX timestamp
+    fn last_accessed(&self) -> u64;
+
+    /// the last time the file was modified in nanoseconds as a UNIX timestamp
+    fn last_modified(&self) -> u64;
+
+    /// the time at which the file was created in nanoseconds as a UNIX timestamp
+    fn created_time(&self) -> u64;
+
+    /// the size of the file in bytes
+    fn size(&self) -> u64;
+
+    /// Change the size of the file, if the `new_size` is greater than the current size
+    /// the extra bytes will be allocated and zeroed
+    fn set_len(&mut self, new_size: u64) -> Result<(), FsError>;
```

Comment:
Usually, this method is named `truncate`. I think it's convey the intend much better. Thought?